### PR TITLE
Roomba controller

### PR DESCRIPTION
**Pull request by asaba96**
Created a mini abstract that handles all of the states of hitting and blocking a roomba. State transitions include:
1) track (either with or without overshoot)
2) hit or block roomba
3) recover height (if failed or succeeded)
4) hold current position for as long as desired (0 seconds means no holding)

Additionally, created test game abstract to test roomba controller

### CODE REVIEW

File: scripts/game.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#! /usr/bin/env python
```
Comment (amiller27) [inline scripts/game.py]: Can we rename this abstract?  Maybe `roomba_controller_test.py`?

File: scripts/game.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    global roomba_array
+    roomba_array = data
+
+def _receive_roomba_controller_status(data):
+    global controller_state
+    controller_state = data
+
+if __name__ == '__main__':
+    try:
+        roomba_array = []
```
Comment (amiller27) [inline scripts/game.py]: You should set this to `None`, not `[]`, because it's not supposed to be an array

File: scripts/game.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    goal = QuadMoveGoal(movement_type="takeoff")
+    # Sends the goal to the action server.
+    client.send_goal(goal)
+    # Waits for the server to finish performing the action.
+    client.wait_for_result()
+    rospy.logwarn("Takeoff success: {}".format(client.get_result()))
+    rospy.sleep(2.0)
+
+    # change element in array to test diff roombas
+    roomba_id = roomba_array.data[5].child_frame_id
```
Comment (amiller27) [inline scripts/game.py]: You need to wait for `roomba_array` to be valid either before this line or before you call `hit_roomba`

File: scripts/game.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    try:
+        roomba_array = []
+        # Initializes a rospy node so that the SimpleActionClient can
+        # publish and subscribe over ROS.
+        _roomba_status_sub = rospy.Subscriber('roombas', 
+                         OdometryArray, _receive_roomba_status)
+
+        rospy.init_node('game')
+        hit_roomba()
+        while not rospy.is_shutdown():
```
Comment (amiller27) [inline scripts/game.py]: You can call `rospy.spin()` instead of using this loop

File: scripts/game.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            pass
+
+    except Exception, e:
```
Comment (amiller27) [inline scripts/game.py]: Probably don't need this whole `try..except` block, I'm guessing you copied it from another abstract?  We should figure out an actual best practice for these

File: scripts/roomba_controller.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+import sys
+import rospy
+from iarc7_msgs.msg import OdometryArray
+import actionlib
+import tf2_ros
+import threading
+from iarc7_motion.msg import QuadMoveGoal, QuadMoveAction
+from iarc7_safety.SafetyClient import SafetyClient
+
+class RoombaRequests(object):
```
Comment (amiller27) [inline scripts/roomba_controller.py]: This is for a single request, no?  Should probably be called `RoombaRequest`

File: scripts/roomba_controller.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    def __init__(self, frame_id, tracking_mode, time_to_hold_once_done):
+        self.frame_id = frame_id
+        self.tracking_mode = tracking_mode
+        self.time_to_hold = time_to_hold_once_done
+
+class RoombaControllerStates(object):
```
Comment (amiller27) [inline scripts/roomba_controller.py]: These should be `ALL_CAPS` (e.g. `TRACKING`), also class name should probably be singular

File: scripts/roomba_controller.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    recover_from_success = 4
+    hold_position = 5
+    completed = 6
+    failed_task = 7
+    failed_recovery = 8
+    invalid_state = 9
+
+class RoombaController(object): 
+
+    def __init__(self, roomba_request, client, status_callback):
```
Comment (amiller27) [inline scripts/roomba_controller.py]: Can you add a docstring explaining what these arguments are?

File: scripts/roomba_controller.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        self._status_callback = status_callback
+        self._client = client
+        roomba_id = roomba_request.frame_id
+        self._roomba_id = roomba_id [0:len(roomba_id)-10]
```
Comment (amiller27) [inline scripts/roomba_controller.py]: No space before the `[` here, might want spaces around the `:` but that's optional

Comment (amiller27) [reply scripts/roomba_controller.py]: Also you need a comment explaining what the `10` is

File: scripts/roomba_controller.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                else: 
+                    rospy.logerr("Roomba Controller is in an invalid state")
+                    self._end = True
+                    self._state = RoombaControllerStates.invalid_state
+
+            self._status_callback(self._state)    
+
+    def run(self):
+        next_thread = threading.Thread(target=self._next_goal)
+        next_thread.setDaemon(True)
```
Comment (amiller27) [inline scripts/roomba_controller.py]: Is there a good reason this is a daemon thread?

File: scripts/roomba_controller.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+        self._is_complete = False
+        self._canceled = False
+        self._end = False
+
+        self._holding = (self._time_to_hold != 0)
+
+        self._state = RoombaControllerStates.tracking
+
+    def _next_goal(self):
```
Comment (amiller27) [inline scripts/roomba_controller.py]: `_next_goal` isn't all that explanatory of what this does

File: scripts/roomba_controller.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    block = 2
+    hit = 3
+    recover_from_success = 4
+    hold_position = 5
+    completed = 6
+    failed_task = 7
+    failed_recovery = 8
+    invalid_state = 9
+
+class RoombaController(object):
```
Comment (amiller27) [inline scripts/roomba_controller.py]: This is more like a `RoombaRequestExecutor`, no?

File: scripts/roomba_controller.py (hunk 8 above)
Comment (amiller27) [reply scripts/roomba_controller.py]: Also are action clients thread safe?

File: scripts/roomba_controller.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            elif self._state == RoombaControllerStates.block:
+                goal = QuadMoveGoal(movement_type="block_roomba", frame_id=self._roomba_id)
+                # Sends the goal to the action server.
+                self._client.send_goal(goal)
+                # Waits for the server to finish performing the action.
+                self._client.wait_for_result()
+                rospy.logwarn("BlockRoomba success: {}".format(self._client.get_result()))
+
+            elif self._state == RoombaControllerStates.hold_position:
+                goal = QuadMoveGoal(movement_type="hold_position", hold_current_position = True)
```
Comment (amiller27) [inline scripts/roomba_controller.py]: No spaces around `=` here

**Comment by amiller27** [comment]
This needs to get rebased

File: scripts/roomba_controller.py (hunk 10 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: I was told to use a daemon thread. so no clue

File: scripts/roomba_controller.py (hunk 13 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: fixed for each one

File: scripts/roomba_controller.py (hunk 11 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: fixed. renamed it to run and run to start.

File: scripts/roomba_controller.py (hunk 9 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: fixed, but keeping : in because it looks cool

File: scripts/roomba_controller.py (hunk 12 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: fixed

File: scripts/roomba_controller.py (hunk 7 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: fixed

File: scripts/roomba_controller.py (hunk 6 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: fixed

File: scripts/game.py (hunk 5 above)
Comment (asaba96) [reply scripts/game.py]: we are choosing to ignore for now

File: scripts/game.py (hunk 4 above)
Comment (asaba96) [reply scripts/game.py]: fixed

File: scripts/roomba_controller.py (hunk 8 above)
Comment (asaba96) [reply scripts/roomba_controller.py]: docstring added; don't know about thread safe

File: scripts/game.py (hunk 2 above)
Comment (asaba96) [reply scripts/game.py]: fixed

File: scripts/game.py (hunk 3 above)
Comment (asaba96) [reply scripts/game.py]: fixed

File: scripts/game.py (hunk 1 above)
Comment (asaba96) [reply scripts/game.py]: fixed

**Pull request closed without merging**